[llvm][LLD][COFF] Add fat-lto-object support for COFF targets

thrimbor opened the pull request:
This adds support for FatLTO to COFF targets in clang and lld.

The changes are adapted from https://github.com/llvm/llvm-project/commit/610fc5cbcc8b68879c562f6458608afe2473ab7f and https://github.com/llvm/llvm-project/commit/14e3bec8fc3e1f10c3dc57277ae3dbf9a4087b1c but much smaller because it just needed the COFF-specific parts wired in, and I tried my best to adapt the pre-existing ELF tests for the COFF version.

My main goal is to be able to use this for shipping pre-built https://github.com/XboxDev/nxdk container images someday, which uses the `i386-pc-win32` target.

tru reviewed (changes requested):
Thanks for working on this. It needs a release note before it can be accepted.

thrimbor commented:
I rebased the PR and added a ReleaseNote entry for LLD based on the one that was added for ELF FatLTO. This is the first time I've done that, so if there's something not quite right with it please let me know.

tru reviewed (approved):
LGTM now, but wait for at least one more reviewer that looks at the code.

thrimbor commented:
Friendly ping for review/merge @mstorsjo @nikic @bd1976bris @aganea @MaskRay

aganea commented on lld/test/COFF/fatlto/fatlto-invalid.s:
```diff
@@ -0,0 +1,8 @@
+# RUN: llvm-mc -filetype=obj -triple=i386-pc-win32 %s -o %t
+# RUN: env LLD_IN_TEST=1 not lld-link %t /out:/dev/null /fat-lto-objects 2>&1 | FileCheck %s
+
```
Are you able to handle a COFF error here as well? For example, when dealing with a FAT LTO .OBJ, any of the `object_error::parse_failed` returned by `COFFObjectFile::initialize`?

mstorsjo commented:
I had a quick readthrough, and didn't find anything objectionable on top of what @aganea already pointed out.

The PR seems to have two separate commits though, that I feel would benefit of staying split into two separate commits in the end. I.e. that would require splitting out at least the Clang change to a separate PR after the LLD change. (The change to EmbedBitcodePass probably can go along with the LLD change, I guess it's required for that to work.)

Then as a small request, I'd appreciate if you could hook up this new option to the LLD MinGW frontend as well. That just requires adding a similar option there (usually it's possible to copy over the option definition from the ELF linker) which just forwards this option to the internal lld-link interface. I guess that'd also be a separate PR aside from this.

thrimbor committed:
[llvm][LLD][COFF] Add fat-lto-object support for COFF targets

thrimbor commented on clang/test/CodeGen/fat-lto-objects.c:
```diff
@@ -67,6 +77,10 @@
 // ASM-NEXT:        .asciz  "BC
 // ASM-NEXT: .size   .Lllvm.embedded.object
 
+//      ASM-COFF: .section .llvm.lto,"ynD"
```
Done, but split off as per @mstorsjo 's recommendation

thrimbor commented on lld/COFF/Driver.cpp:
```diff
@@ -256,6 +257,23 @@ MemoryBufferRef LinkerDriver::takeBuffer(std::unique_ptr<MemoryBuffer> mb) {
   return mbref;
 }
 
+InputFile *LinkerDriver::tryCreateFatLTOFile(MemoryBufferRef mb,
```
Done

thrimbor commented:
I've now tried to address all the review comments, split off the clang commit and rebased the branch

mstorsjo reviewed (approved):
LGTM. But I'd still want to wait to see if @aganea has any further concerns.

mstorsjo commented:
@thrimbor FWIW, please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account. See [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#email-addresses) and [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

mstorsjo commented:
I pushed 421b080d189e379766db1ffed9d6e08aa470f78c to fix the failing buildbots.
